lib/ble.toit

```diff
@@ -517,134 +512,381 @@ class Device:
-    resource_state_.clear_state CONNECTED_EVENT_
+/**
+The manager for advertising and managing local services.
+*/
 
-  wait_for_client_disconnected -> none:
-    resource_state_.wait_for_state DISCONNECTED_EVENT_
-    resource_state_.clear_state DISCONNECTED_EVENT_
+class PeripheralManager extends Resource_:
+  static DEFAULT_INTERVAL ::= Duration --us=46875
+  adapter/Adapter
 
-STARTED_EVENT_              ::= 1 << 0
-COMPLETED_EVENT_            ::= 1 << 1
-DISCOVERY_EVENT_            ::= 1 << 2
-CONNECTED_EVENT_            ::= 1 << 3
-CONNECT_FAILED_EVENT_       ::= 1 << 4
-DISCONNECTED_EVENT_         ::= 1 << 5
+  constructor .adapter:
+    resource := ble_create_peripheral_manager_ adapter.resource_group_
+    super adapter.resource_group_ resource --auto_release
+    resource_state_.wait_for_state STARTED_EVENT_
 
-ble_set_preferred_mtu_ mtu:
-  #primitive.ble.set_preferred_mtu
+  /**
+  Starts advertising the $data.
+
+  The data is advertised once every $interval.
 
-ble_init_ config_resource_group:
+  The advertise will include the given $connection_mode, use one
+    of the BLE_CONNECTION_MODE_* constants.
+
+  If the adapter does not support parts of the advertise content, INVALID_ARGUMENT is thrown.
+  For example, on MacOS manufacturing data can not be specified.
+
+  If the adapter does not allow configuration of $interval or $connection_mode an INVALID_ARGUMENT is thrown.
+  */
+  start_advertise
+      data/AdvertisementData
+      --interval/Duration=DEFAULT_INTERVAL
+      --connection_mode/int=BLE_CONNECT_MODE_NONE:
+    if platform == PLATFORM_MACOS:
+      if interval != DEFAULT_INTERVAL or connection_mode != BLE_CONNECT_MODE_NONE: throw "INVALID_ARGUMENT"
+
+    raw_service_classes := Array_ data.service_classes.size null
+
+    data.service_classes.size.repeat:
+      id/BleUUID := data.service_classes[it]
+      raw_service_classes[it] = id.encode_for_platform_
+    ble_advertise_start_
+      resource_
+      data.name or ""
+      raw_service_classes
+      data.manufacturer_data
+      interval.in_us
+      connection_mode
+
+    state := resource_state_.wait_for_state ADVERTISE_START_SUCEEDED_EVENT_ | ADVERTISE_START_FAILED_EVENT_
+    if state & ADVERTISE_START_FAILED_EVENT_ != 0: throw "Failed to start advertising"
+
+  /**
+    Stops advertising.
+  */
+  stop_advertise:
+    ble_advertise_stop_ resource_
+
+  /**
+    Adds a new service to the peripheral identified by $uuid. The returned service should be configured with
+    the appropriate characteristics and then be deployed.
+  */
+  add_service uuid/BleUUID -> LocalService:
+    return LocalService this uuid
+
+
+class AdapterInformation:
+  identifier/string
+  address/ByteArray
+  supports_central_role/bool
+  supports_peripheral_role/bool
+  handle_/any
+  constructor .identifier .address .supports_central_role .supports_peripheral_role .handle_:
+
+class Adapter:
+  static adapters -> List/*<AdapterInfo>*/:
```

Comment:
There are some side effects, in that the platform actually initializes the adapter in the constructor. I believe that is why I kept them separate. So if you have multiple adapters, there might be other processes/programs that needs the adapter and this would cause issues. I will rename the name. Also, only relevant on windows and linux, that isn't implemented yet.